src/java.base/share/classes/sun/security/ssl/SSLSessionImpl.java

```diff
@@ -1486,6 +1605,214 @@ public List<SNIServerName> getRequestedServerNames() {
+            }
+
+            // Perform RFC 5705 calculations using the internal SunJCE PRF.
+            String prfAlg;
+            HashAlg hashAlg;
+            if (protocolVersion == TLS12) {
+                prfAlg = "SunTls12Prf";
+                hashAlg = cipherSuite.hashAlg;
+            } else {  //  all other cases
+                prfAlg = "SunTlsPrf";
+                hashAlg = H_NONE;
+            }
+
+            // Make a seed with randoms and optional context
+            // Note that if context is null, it is omitted from the calc
+            byte[] clientRandomBytes = clientRandom.randomBytes;
+            byte[] serverRandomBytes = serverRandom.randomBytes;
+            byte[] seed = new byte[
+                    clientRandomBytes.length + serverRandomBytes.length +
+                            ((context != null) ? (2 + context.length) : 0)];
+
+            int pos = 0;
+            System.arraycopy(
+                    clientRandomBytes, 0, seed, pos, clientRandomBytes.length);
+            pos += clientRandomBytes.length;
+            System.arraycopy(
+                    serverRandomBytes, 0, seed, pos, serverRandomBytes.length);
+            pos += serverRandomBytes.length;
+            if (context != null) {
+                // RFC 5705, "If no context is provided, ..."
+                seed[pos++] = (byte) ((context.length >> 8) & 0xFF);
+                seed[pos++] = (byte) ((context.length) & 0xFF);
+                System.arraycopy(
+                        context, 0, seed, pos, context.length);
+            }
+
+            // Call the PRF function.
+            try {
+                @SuppressWarnings("deprecation")
+                TlsPrfParameterSpec spec = new TlsPrfParameterSpec(
+                        masterSecret, label, seed, length,
+                        hashAlg.name, hashAlg.hashLength, hashAlg.blockSize);
+                KeyGenerator kg = KeyGenerator.getInstance(prfAlg);
+                kg.init(spec);
+                return kg.generateKey();
+            } catch (NoSuchAlgorithmException |
+                     InvalidAlgorithmParameterException e) {
+                throw new SSLKeyException("Could not generate Exporter/PRF", e);
+            }
+        default:
+            // SSLv3 is vulnerable to a triple handshake attack and can't be
+            // mitigated by RFC 7627.  Don't support this or any other
+            // unknown protocol.
+            throw new SSLKeyException(
+                    "Exporters not supported in " + protocolVersion);
+        }
+    }
+
+    /**
+     * Generate Exported Key Material (EKM) calculated according to the
+     * algorithms defined in RFCs 5705/8446.
+     */
+    @Override
+    public byte[] exportKeyingMaterialData(
+            String label, byte[] context, int length) throws SSLKeyException {
+        byte[] bytes =
+                exportKeyingMaterialKey(label, context, length).getEncoded();
```

Comment:
In PKCS #11, calling `deriveKey(...).getEncoded()` is not equivalent to `deriveData()`. It's quite likely that `deriveKey` returns an un-extractable key, but `deriveData` using the exact same input returns the keying material.

Reply:
Adjusted for the comment.